### PR TITLE
win_secedit module: adds ability to mod local security policies

### PR DESCRIPTION
**Pull request by rndmh3ro**
##### ISSUE TYPE

Feature Pull Request
##### COMPONENT NAME

win_secedit
##### ANSIBLE VERSION

```
ansible 2.2.0.0
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

Allows users to modify local security policies via the secedit utility for example:

``` yml
    - name: Set local password age policy to 14 days
      win_secedit:
        category: System Access
        key: MaximumPasswordAge
        value: 14
```

### CODE REVIEW

**Comment by rndmh3ro** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/3214 by rndmh3ro (not original author)

File: lib/ansible/modules/windows/win_secedit.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+short_description: Manipulate local security policies via secedit
+description:
+    - Modifies key values for local security policies via secedit
+options:
+  category:
+    description:
+      - The category you wish to modify a value under. This can things like System Access, Event Audit, etc.
+        If you supply an invalid category the module will error out and let you know what the valid categories are for that particular system.
+    required: true
+    default: null
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_secedit.py:41]: If parameter is required, there is no default.

File: lib/ansible/modules/windows/win_secedit.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  key:
+    description:
+      - The key under the category for which you wish to modify the value.
+        For example, under the System Access category there is a key MinimumPasswordAge that could be targeted.
+        Just like with category, if an invalid key is specified, the module will error out and show what the valid keys for the given category are.
+    required: true
+    default: null
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_secedit.py:48]: Same.

File: lib/ansible/modules/windows/win_secedit.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  value:
+    description:
+      - The value to assign to the key.
+    required: true
+    default: null
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_secedit.py:53]: Same.

**Comment by jborean93** [comment]
Hey @rndmh3ro and @defionscode I'll try and get to looking at this sometime soon, looks like a good module to have, I know I've written a once off script to do this before. There are a few things in there from a brief look that would need to be made more standard to our other Windows module and I'll add some comments for a first round look.

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# WANT_JSON
+# POWERSHELL_COMMON
+
+########
+
+
+Set-StrictMode -Version Latest
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:25]: We usually set StrictMode to 2 when running a module automatically, this should be needed

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+function Get-IniFile {
+    param (
+        [parameter(mandatory=$true, position=0, valuefrompipelinebypropertyname=$true, valuefrompipeline=$true)][string]$FilePath
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:29]: Are we able to just move this to function Get-IniFile($FilePath) to make it more python like

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+        $ini = New-Object System.Collections.Specialized.OrderedDictionary
+        $currentSection = New-Object System.Collections.Specialized.OrderedDictionary
+        $curSectionName = "default"
+
+        switch -regex (gc $FilePath)
+        {
+            "^\[(?<Section>.*)\]"
+            {
+                        $ini.Add($curSectionName, $currentSection)
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:40]: Probably want to remove an indent for these lines so they match up in each case statement

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                        # add to current section Hash Set
+                        $currentSection.Add($Matches['Key'].trim(), $Matches['Value'].trim())
+                }
+                "^$"
+                {
+                        # ignore blank line
+                }
+                default
+                {
+                        throw "Unidentified: $_"  # should not happen
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:56]: Change to Fail-Json @{} "Error Message here" and change the error to be a bit more descriptive

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                }
+        }
+        if ($ini.Keys -notcontains $curSectionName) { $ini.Add($curSectionName, $currentSection) }
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:59]: Move the if body to a new line

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+        return $ini
+}
+
+function Out-IniFile{
+    param (
+        [parameter(mandatory=$true, position=0, valuefrompipelinebypropertyname=$true, valuefrompipeline=$true)][System.Collections.Specialized.OrderedDictionary]$ini,
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:66]: Probably want more input from the other Windows guys but I don't think we want to do pipelining in the modules if we can avoid it. I believe this is so we keep a similar standard between the python and powershell ones. @jhawkesworth @dagwieers @nitzmahone?

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        }
+        else
+        {
+                $output | Out-File -FilePath $FilePath -Encoding:ASCII
+        }
+}
+
+$params = Parse-Args $args;
+
+$result = New-Object psobject @{
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:100]: `$result = @{`

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    changed = $false
+};
+
+$category = Get-Attr $params "category" -failifempty $true
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:104]: Have a look at `Get-AnsibleParam` which can be seen in other modules like `win_tempfile`. This way you can specify the object type and other options.

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+$key = Get-Attr $params "key" -failifempty $true
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:105]: Change to `Get-AnsibleParam` as above

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+$value = Get-Attr $params "value" -failifempty $true
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:106]: Change to `Get-AnsibleParam` as above

File: lib/ansible/modules/windows/win_secedit.ps1 [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+$sepath = "$home\sec_edit_dump.inf"
+
+If ((Get-WmiObject Win32_ComputerSystem).PartOfDomain) {
+    Fail-Json $result "This host is joined to a Domain Controller, you'll need to modify GPO directly instead of secedit"
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_secedit.ps1:110]: I don't think this should be here, in some GPO policies I've seen set definitely allowed me to change certain entries just not others. This should probably be a warning in the python document.

**Review by jborean93 — CHANGES_REQUESTED** [review]
This is probably enough comments for a first round check. Three major things that would need to be done would be to;
* Add check mode to the module
* Add some tests
* More of a question for us to answer in the next WUG around how do we want to handle pipelining and indenting in powershell modules. I'll raise an agenda item.

Please feel free to ask me questions around any of my comments or if you need guidance around check mode/testing.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by jborean93** [comment]
@rndmh3ro are you still willing to look into getting this merged in. I'm happy to take it on if you don't have the time as I think this would be good to have in Ansible.

**Comment by rndmh3ro** [comment]
@jborean93, feel free to take this on!

**Comment by jhawkesworth** [comment]
@jborean93 - sorry I couldn't make the WWG meeting.
I'd prefer to avoid pipelining in ansible powershell modules just because I think its another thing to understand.
I think splitting into 2 modules is valid - perhaps win_user_privilege and win_audit_event - I think the first one would be much more useful than the second one.  I'm less concerned about common code at the moment - eventually we will be able to share code between modules - but there's also the issue of 'findability'.  I don't think we have the ability to do a redirect in the documentation, but perhaps we could have a 'stub' win_secedit or win_secpol module which just has documentation links to win_user_privilege and win_audit_event.  If stub modules aren't acceptable, maybe just mentioning secpol and 'security policy' in the short module description is enough.

**Comment by jborean93** [comment]
No worries time zones can be quite annoying, the general consensus was to split it up into separate modules mostly focusing on the user right stuff and one to do what this PR focuses on. This is similar to how win_path and win_environment fit together.

**Comment by juliedavila** [comment]
Given where this has gone, should we close this particular pr?

**Comment by rndmh3ro** [comment]
Yes, #26332 superseeds this.

**Pull request closed without merging**